### PR TITLE
[UI-60] Tree component

### PR DESCRIPTION
**Pull request by PiotrGrzechnik**
#### Task

- **Issue or task referred:** UI-60
- **Feature or Bugfix:** Feature

Tree component which display treeview with nested nodes. User can select each node. InitialOpen prop define whether Tree should be open on load or not. Also user can choose smoothness of collapse.

#### Checklist

- [x] Required unit tests prepared
- [x] Documentation prepared
- [x] External licenses validated (ideally MIT)

### CODE REVIEW

File: packages/tree/src/Tree.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+import { buildClassName } from '@talixo/shared'
+
+import TreeNode from './TreeNode'
+
+const moduleName = 'tree'
+
+/**
+ * Function to recursivly define propTypes
+ * @param {func} f
```
Comment (maciek15) [inline packages/tree/src/Tree.js]: Both here should be `{function}`

File: packages/tree/src/Tree.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  initialOpen: false,
+  selectEnabled: false,
+  smooth: true
+}
+/**
+ * Component which represents Tree.
+ *
+ * @param {object} props
+ * @param {string} [props.className]
+ * @param {array} [props.data]
```
Comment (maciek15) [inline packages/tree/src/Tree.js]: If sth is required you should omit brackets

File: packages/tree/stories.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+addStory('initial', readme, () => (
+  <Tree data={[{id: 1,
+    name: 'animal',
+    children:
+    [{id: 2, name: 'tiger'},
+      {id: 3, name: 'cow'}]},
+  {id: 4, name: 'people'}]} />
+))
+
+addStory('more nesting', readme, () => (
```
Comment (maciek15) [inline packages/tree/stories.js]: Nested maybe?

File: packages/tree/src/TreeNode.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    })
+  }
+
+  render () {
+    const { children, initialOpen, node, selectEnabled, smooth } = this.props
+    const { collapsed, selected } = this.state
+    const nodeCls = buildClassName([moduleName, 'node'], null, { selected, childless: !children })
+    const nodeNameCls = buildClassName([moduleName, 'node-name'])
+    const childrenCls = buildClassName([moduleName, 'node-children'])
+    const iconCls = buildClassName([moduleName, 'node-icon'])
```
Comment (maciek15) [inline packages/tree/src/TreeNode.js]: It should be: `buildClassName([moduleName, 'node', 'icon'])` which gives you correct class: `${moduleName}__node__icon`. Same thing with children and name.

File: packages/tree/src/TreeNode.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            children={el.children}
+            initialOpen={initialOpen}
+            key={i}
+            node={el}
+            selectEnabled={selectEnabled}
+            smooth={smooth}
+          />
+        )
+      })
+    }
```
Comment (maciek15) [inline packages/tree/src/TreeNode.js]: Put this into separate function (i.e. generate children) inside class

File: packages/tree/src/TreeNode.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  state = {
+    selected: false,
+    collapsed: !this.props.initialOpen
+  }
+
+  /**
+   * Function which handle click on node, change select
+   * @param {*} props
+   */
+  handleClick = (props) => {
```
Comment (maciek15) [inline packages/tree/src/TreeNode.js]: It does not handle any user defined events like onClick or onSelect from props. In fact selecting item only change its color.

File: packages/tree/src/TreeNode.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    const iconCls = buildClassName([moduleName, 'node-icon'])
+    const icon = this.state.collapsed ? 'chevron_right' : 'expand_more'
+    let iconContainer
+    let nodes
+
+    if (children) {
+      iconContainer = <span
+        className={iconCls}
+        onClick={this.toggle} >
+        {<Icon name={`${icon}`} />}
```
Comment (maciek15) [inline packages/tree/src/TreeNode.js]: Or simply `<Icon name={icon} />`

File: packages/tree/src/TreeNode.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+import { buildClassName } from '@talixo/shared'
+import { Collapse } from '@talixo/collapse'
+import { Icon } from '@talixo/icon'
+
+const moduleName = 'tree'
+
+/**
+ * Function to recursivly define propTypes
+ * @param {func} f
+ */
```
Comment (maciek15) [inline packages/tree/src/TreeNode.js]: It also returns function

File: packages/tree/styles/main.sass [hunk 9]
```diff
@@ -0,0 +1,37 @@
+@import './config'
```
Comment (maciek15) [inline packages/tree/styles/main.sass]: In other components this kind of file is called `geometry.sass` and is imported to inside `main.sass`

File: packages/tree/styles/main.sass [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    margin-bottom: dp(0.2)
+
+    &-name
+      display: inline-block;
+      margin-bottom: dp(0.3)
+
+    &-children
+      list-style-type: none
+
+    &-icon
```
Comment (maciek15) [inline packages/tree/styles/main.sass]: Add here `user-select: none`. It prevents caret from being selected

File: packages/tree/styles/main.sass [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    box-sizing: border-box
+    display: inline-block
+    border-radius: dp(0.4)
+    width: 100%
+    padding: dp(0.9)
+    margin-top: dp(0.2)
+    margin-bottom: dp(0.2)
+
+    &-name
+      display: inline-block;
```
Comment (maciek15) [inline packages/tree/styles/main.sass]: Remove semi-colon

File: packages/tree/styles/main.sass [hunk 12]
```diff
@@ -0,0 +1,37 @@
+@import './config'
+
+@include module('tree')
+  display: inline-block
+  cursor: pointer
```
Comment (maciek15) [inline packages/tree/styles/main.sass]: Cursor pointer should indicate that something will happen after clicking given item. So apply it only on caret (it opens tree) and on item items which are selectable (when `selectEnabled = true`) not on whole component

File: packages/tree/styles/main.sass [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    display: inline-block
+    border-radius: dp(0.4)
+    width: 100%
+    padding: dp(0.9)
+    margin-top: dp(0.2)
+    margin-bottom: dp(0.2)
+
+    &-name
+      display: inline-block;
+      margin-bottom: dp(0.3)
```
Comment (maciek15) [inline packages/tree/styles/main.sass]: There was something inside previous commits comment that we should use iteration of 0.125. So here you can change to .375. Same with other dps in this file

File: packages/tree/README.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+Property name | Type      | Default | Description                    
+--------------|-----------|:-------:|--------------------------------
+className     | string    | n/a     | Additional class name passed to wrapper
+data          | DataType[]| n/a     | Data passed to component
+initialOpen   | bool      | `false` | Open tree on load
+selectEnabled | bool      | `false` | Enable to select each node
+smooth        | bool      | `true`  | Animation of collapse
+
+## DataType
```
Comment (maciek15) [inline packages/tree/README.md]: This should be 
\## Property shapes

\### DataType

File: packages/tree/src/TreeNode.js [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+      children.map((el, i) => {
+        return (
+          <TreeNode
+            children={el.children}
+            initiallyOpen={initiallyOpen}
+            key={i}
+            node={el}
+            smooth={smooth}
+            onClick={onClick}
+          />
```
Comment (maciek15) [inline packages/tree/src/TreeNode.js]: Missing `cursor: pointer` when element is clickable

File: packages/tree/src/TreeNode.js [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    const nodeCls = buildClassName([moduleName, 'node'], null, { selected, childless: !children })
+    const nodeNameCls = buildClassName([moduleName, 'node-name'], null, {'selectable': onClick})
+    const childrenCls = buildClassName([moduleName, 'node-children'])
+
+    return (
+      <span>
+        <li className={nodeCls}>
+          {children && this.renderIcon()}
+          <span className={nodeNameCls} onClick={this.handleClick}>
+            {node && (node.render && typeof node.render ? node.render(node.name) : node.name) }
```
Comment (rangoo94) [inline packages/tree/src/TreeNode.js]: It should have `render` passed instead

File: packages/tree/src/TreeNode.js [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+  render () {
+    const { children, node, onClick, smooth } = this.props
+    const { collapsed, selected } = this.state
+    const nodeCls = buildClassName([moduleName, 'node'], null, { selected, childless: !children })
+    const nodeNameCls = buildClassName([moduleName, 'node-name'], null, {'selectable': onClick})
+    const childrenCls = buildClassName([moduleName, 'node-children'])
+
+    return (
+      <span>
+        <li className={nodeCls}>
```
Comment (rangoo94) [inline packages/tree/src/TreeNode.js]: `li` inside `span`?

File: packages/tree/README.md [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+## Property shapes
+
+## DataType
+
+Type name       | Type              |
+----------------|----------------------------------
+id              |  number           | id of node
+name            |  string           | name od node
+children        |  dataType[]       | children of node
```
Comment (rangoo94) [inline packages/tree/README.md]: what is `dataType`?

File: packages/tree/src/Tree.js [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  const { data, initiallyOpen, smooth, ...restProps } = props
+  const clsName = buildClassName(moduleName)
+  const children = data.map((el, i) =>
+    <TreeNode
+      initiallyOpen={initiallyOpen}
+      key={i}
+      node={el}
+      children={el.children}
+      smooth={smooth}
+      {...restProps}
```
Comment (rangoo94) [inline packages/tree/src/Tree.js]: It should be passed to `ul` instead

File: packages/tree/src/Tree.js [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+import TreeNode from './TreeNode'
+
+const moduleName = 'tree'
+
+/**
+ * Function to recursivly define propTypes
+ * @param {function} f
+ * @returns {function}
+ */
+function lazyFunction (f) {
```
Comment (rangoo94) [inline packages/tree/src/Tree.js]: What is that for?

**Comment by rangoo94** [comment]
You can't control here a tree from outside.

**Comment by rangoo94** [comment]
I just made some changes to tree, but still it needs:
* better styling
* control from outside + correctly self-controlled (both if it goes about collapsing and selection)
* toggle shouldn't be only on arrow
* `buildId` similar to the one in `Table` or `Accordion`
* updated documentation, stories & unit tests

**Comment by rangoo94** [comment]
I am closing this ticket in favor of unified navigation which @maciek15 will create. Thanks anyway!

**Pull request closed without merging**